ext/debug/ctf/bottoms/posix/ctf_bottom.h

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (c) 2018 Oticon A/S
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef _CTF_BOTTOM_POSIX_H
+#define _CTF_BOTTOM_POSIX_H
+
+#include <stddef.h>
+#include <string.h>
+#include <stdio.h>
+#include <zephyr/types.h>
+#include <ctf_map.h>
+
+
+/* Obtain a field's size at compile-time.
+ * Internal to this bottom-layer.
+ */
+#define CTF_BOTTOM_INTERNAL_FIELD_SIZE(x)      + sizeof(x)
+
+/* Append a field to current event-packet.
+ * Internal to this bottom-layer.
+ */
+#define CTF_BOTTOM_INTERNAL_FIELD_APPEND(x)		 \
+	{						 \
+		memcpy(epacket_cursor, &(x), sizeof(x)); \
+		epacket_cursor += sizeof(x);		 \
+	}
+
+/* Gather fields to a contiguous event-packet, then atomically emit.
+ * Used by middle-layer.
+ */
+#define CTF_BOTTOM_FIELDS(...)						     \
+	{								     \
+		u8_t epacket[						     \
+			0 MAP(CTF_BOTTOM_INTERNAL_FIELD_SIZE, ##__VA_ARGS__) \
+		];							     \
+		u8_t *epacket_cursor = &epacket[0];			     \
+									     \
+		MAP(CTF_BOTTOM_INTERNAL_FIELD_APPEND, ##__VA_ARGS__)	     \
+		ctf_bottom_emit(epacket, sizeof(epacket));		     \
+	}
+
+/* No need for locking when ctf_bottom_emit does POSIX fwrite(3) which is thread safe.
+ * Used by middle-layer.
+ */
+#define CTF_BOTTOM_LOCK         { /* empty */ }
```

Comment:
Could we use function-like invocation of macros that could execute code?
`CFT_BOTTOM_LOCK(); CTF_BOTTOM_UNLOCK();`

Reply:
Absolutely! If it makes it more readable then I'm all for it.